provider/src/main/kotlin/org/gradle/kotlin/dsl/ExtraPropertiesExtensions.kt

```diff
@@ -35,18 +35,51 @@ val ExtensionAware.extra: ExtraPropertiesExtension
     get() = extensions.extraProperties
 
 
-operator fun <T> ExtraPropertiesExtension.setValue(receiver: Any?, property: KProperty<*>, value: T) =
-    set(property.name, value)
+operator fun ExtraPropertiesExtension.provideDelegate(any: Any?, property: KProperty<*>) =
+    if (property.returnType.isMarkedNullable) NullableExtraPropertyDelegate(this, property.name)
+    else NonNullExtraPropertyDelegate(this, property.name)
 
 
-operator fun <T> ExtraPropertiesExtension.getValue(receiver: Any?, property: KProperty<*>): T =
-    property.run {
-        val isFound = has(name)
-        val foundValue = if (isFound) get(name) else null
-        if (returnType.isMarkedNullable || (isFound && foundValue != null)) return uncheckedCast(foundValue)
+/**
+ * Enables typed access to extra properties.
+ */
+interface ExtraPropertyDelegate {
+    operator fun <T> getValue(any: Any?, property: KProperty<*>): T
```

Comment:
:o: Rename `any: Any?` to `receiver` for consistency?

Reply:
👍 done

Reply:
:o: Maybe unify `PropertyDelegate` and `ExtraPropertyDelegate` as `PropertyDelegate` and `MutablePropertyDelegate : PropertyDelegate`?

Reply:
Indeed!
Done ✅